concordium-std/src/test_infrastructure.rs

```diff
@@ -695,6 +649,223 @@ impl<T: AsRef<[u8]>> Seek for ContractStateTest<T> {
+            move |_parameter: Parameter,
+                  _amount: Amount,
+                  _state: &mut State,
+                  _output: &mut Vec<u8>| { Err(error.clone()) },
+        );
+        Self {
+            mock_fn,
+        }
+    }
+}
+
+pub struct HostTest<State> {
+    mocking_fns:      HashMap<(ContractAddress, OwnedEntrypointName), MockFn<State>>,
+    transfers:        Vec<(AccountAddress, Amount)>,
+    contract_balance: Option<Amount>,
+    state:            State,
+    missing_accounts: Vec<AccountAddress>,
+}
+
+impl<State> HasHost<State> for HostTest<State> {
+    type CallResponseType = Cursor<Vec<u8>>;
+
+    /// Perform a transfer to the given account if the contract has sufficient
+    /// balance. Use `make_account_missing` to test out transfers to
+    /// accounts not on chain.
+    ///
+    /// Possible errors:
+    ///   - `InvokeError::AmountTooLarge`: Contract has insufficient funds.
+    ///   - `InvokeError::MissingAccount`: Attempted transfer to an account set
+    ///     a missing with `make_account_missing`.
+    fn invoke_transfer(&mut self, receiver: &AccountAddress, amount: Amount) -> InvokeResult<()> {
+        if self.missing_accounts.contains(receiver) {
+            return Err(InvokeError::MissingAccount);
+        }
+        let contract_balance = unwrap_contract_balance(&mut self.contract_balance);
+        if *contract_balance >= amount {
+            *contract_balance -= amount;
+            self.transfers.push((receiver.clone(), amount));
+            Ok(())
+        } else {
+            Err(InvokeError::AmountTooLarge)
+        }
+    }
+
+    /// Invoke a contract.
+    fn invoke_contract(
+        &mut self,
+        to: &ContractAddress,
+        parameter: Parameter,
+        method: EntrypointName,
+        amount: Amount,
+    ) -> InvokeResult<(bool, Option<Self::CallResponseType>)> {
+        let handler = match self.mocking_fns.get(&(*to, OwnedEntrypointName::from(method))) {
+            Some(handler) => handler,
+            None => fail!(
+                "Mocking has not been set up for invoking contract {:?} with method '{}'.",
+                to,
+                method
+            ),
+        };
+        // Check if the contract has sufficient balance.
+        // Do not try to unwrap the balance if amount is zero.
+        if amount.micro_ccd > 0 && *unwrap_contract_balance(&mut self.contract_balance) < amount {
+            return Err(InvokeError::AmountTooLarge);
+        }
+        let mut output = Vec::new();
+        let res = (handler.mock_fn)(parameter, amount, &mut self.state, &mut output)
+            .and_then(|state_modified| Ok((state_modified, Some(Cursor::new(output)))));
+        if res.is_ok() && amount.micro_ccd > 0 {
+            *unwrap_contract_balance(&mut self.contract_balance) -= amount;
+        }
+        res
+    }
+
+    fn state(&mut self) -> &mut State { &mut self.state }
+
+    fn self_balance(&self) -> Amount {
```

Comment:
Docs

Reply:
Add comment about why this doesn't use unwrap_contract_balance (not possible due to immutable ref to self)